contracts/USM.sol

```diff
@@ -1,117 +1,131 @@
-    uint constant MINT_FEE = UNIT / 1000; // 0.1%
-    uint constant BURN_FEE = UNIT / 200; // 0.5%
+    FUM public fum;
+    uint public latestFumPrice;
 
-    address oracle;
-    uint public ethPool;
+    event LatestFumPriceChanged(uint previous, uint latest);
 
     /**
      * @param _oracle Address of the oracle
      */
-    constructor(address _oracle) public ERC20("Minimal USD", "USM") {
-        oracle = _oracle;
-        ethPool = 0;
-        console.log("Hi!");
-        console.log(_oracle);
+    constructor(address _oracle) public BufferedToken(_oracle, "Minimal USD", "USM") {
+        _setLatestFumPrice(MIN_ETH_AMOUNT);
+        fum = new FUM();
     }
 
     /**
-     * @notice Multiplies x and y, assuming they are both fixed point with 27 digits.
+     * @notice Calculates the price of FUM using its total supply
+     * and ETH buffer
      */
-    function mulFixed(uint256 x, uint256 y) internal pure returns (uint256) {
-        return x.mul(y).div(UNIT);
-    }
+    function fumPrice() public view returns (uint) {
+        int buffer = ethBuffer();
+        // if we're underwater, use the last fum price
+        if (buffer <= 0 || debtRatio() > MAX_DEBT_RATIO) {
+            return latestFumPrice;
+        }
 
-    /**
-     * @notice Divides x by y, assuming they are both fixed point with 27 digits.
-     */
-    function divFixed(uint256 x, uint256 y) internal pure returns (uint256) {
-        return x.mul(UNIT).div(y);
+        uint fumTotalSupply = fum.totalSupply().max(WAD); // fumTotalSupply is floored at WAD
+        return uint(buffer).wadDiv(fumTotalSupply);
     }
 
     /**
-     * @notice Mint ETH for USM. Uses msg.value as the ETH deposit.
+     * @notice Mint ETH for USM with checks and asset transfers. Uses msg.value as the ETH deposit.
      */
     function mint() external payable {
-        uint usmAmount = _ethToUsm(msg.value);
-        uint usmMinusFee = usmAmount.sub(mulFixed(usmAmount, MINT_FEE));
-        ethPool = ethPool.add(msg.value);
-        _mint(msg.sender, usmMinusFee);
+        require(msg.value > MIN_ETH_AMOUNT, "Must deposit more than 0.001 ETH");
+        _mint(msg.value);
+        // set latest fum price
+        _setLatestFumPrice(fumPrice());
     }
 
     /**
-     * @notice Burn USM for ETH.
+     * @notice Burn USM for ETH with checks and asset transfers.
      *
-     * @param _usmAmount Amount of USM to burn.
+     * @param _usmToBurn Amount of USM to burn.
      */
-    function burn(uint _usmAmount) external {
-        uint ethAmount = _usmToEth(_usmAmount);
-        uint ethMinusFee = ethAmount.sub(mulFixed(ethAmount, BURN_FEE));
-        ethPool = ethPool.sub(ethMinusFee);
-        _burn(msg.sender, _usmAmount);
-        Address.sendValue(msg.sender, ethMinusFee);
+    function burn(uint _usmToBurn) external {
+        require(_usmToBurn >= MIN_BURN_AMOUNT, "Must burn at least 1 USM");
+        uint ethToSend = _burn(_usmToBurn);
+        require(debtRatio() <= MAX_DEBT_RATIO,
```

Comment:
I suggest `require(debtRatio() <= WAD)`, ie, 100% rather than `MAX_DEBT_RATIO` = (eg) 80%; because when we're between 80% and 100%, `burns` actually _decrease_ debt ratio!  Eg, suppose price is $400, the pool contains 100 ETH (total value: $40,000), and there are 36,000 USM outstanding, so debt ratio is $36,000 / $40,000 = 90%.  And now suppose a user burns 1,000 USM, removing 2.5 ETH.  Then the new debt ratio is $35,000 / $39,000 = 89.74%.

Reply:
Aaaahhhh 😲  this is interesting. Which functions increase, and which functions decrease `debtRatio`?

Reply:
Fixed :+1:

Reply:
The easy ones are: `fund` decreases debt ratio, `defund` increases it.

The less easy ones: `mint` moves dr **closer to 100%** (so, increases it if <100%, decreases it if >100%), `burn` moves it **further from 100%** (the reverse).  This takes a bit of getting your head around but the main point is that (fees aside) **neither of them changes the size of the buffer.**

One way to think of it is, you can think of the ETH pool as made of two parts: the "covered" portion required to redeem all outstanding USM, and the "uncovered" buffer.  When dr > 100%, the buffer is negative (the system is underwater).  Two `mint` examples show the difference:

1. 20 ETH in pool = 10 ETH covered + 10 ETH buffer: dr = 50%.  `minting` 80 ETH means now 100 in pool = 90 covered + 10 buffer: dr = 90%.  So dr increased.
2. 20 in pool = 30 covered + -10 buffer: dr = 150%.  `minting` 80 means now 100 in pool = 110 covered + -10 buffer: dr = 110%.  So dr decreased.

And the `burn` examples are just the same thing played backwards:

3. 100 in pool = 90 covered + 10 buffer: dr = 90%.  `burning` 80 means now 20 in pool = 10 covered + 10 buffer: dr = 50%.  So dr decreased.
4. 100 in pool = 110 covered + -10 buffer: dr = 110%.  `burning` 80 means now 20 in pool = 30 covered + -10 buffer: dr = 150%.  So dr increased.

Reply:
It is worth wrapping your heads around this because I anticipate you guys presenting about it at a conference sometime :)